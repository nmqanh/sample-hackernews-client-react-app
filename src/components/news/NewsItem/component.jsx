import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { List } from 'antd';
import {
  MessageOutlined,
  ClockCircleOutlined,
  StarOutlined,
  UserOutlined,
} from '@ant-design/icons';
import IconText from '../../shared/IconText';

const HN_NEWS_URL = 'https://news.ycombinator.com/item?id=';
const HN_USER_URL = 'https://news.ycombinator.com/user?id=';

const NewsItem = ({ item }) => (
  <List.Item
    key={item.title}
    actions={[
      <IconText href={HN_NEWS_URL + item.objectID} icon={StarOutlined} text={(item.points).toString()} key="list-vertical-star-o" />,
      <IconText href={HN_USER_URL + item.author} icon={UserOutlined} text={item.author} key="list-vertical-like-o" />,
      <IconText href={HN_NEWS_URL + item.objectID} icon={ClockCircleOutlined} text={moment(item.created_at).fromNow()} key="list-vertical-time-o" />,
      <IconText href={HN_NEWS_URL + item.objectID} icon={MessageOutlined} text={item.num_comments.toString()} key="list-vertical-message" />,
    ]}
  >
    <List.Item.Meta
      title={<a target="blank" href={item.url}>{item.title}</a>}
      description={<a target="blank" href={item.url}>{item.url}</a>}
    />
  </List.Item>
);

export const ItemPropType = PropTypes.shape({
  objectID: PropTypes.string.isRequired,
  title: PropTypes.string,
  url: PropTypes.string,
  points: PropTypes.number,
  num_comments: PropTypes.number,
  author: PropTypes.string,
  created_at: PropTypes.string,
});

NewsItem.propTypes = {
  item: ItemPropType.isRequired,
};

export default NewsItem;
