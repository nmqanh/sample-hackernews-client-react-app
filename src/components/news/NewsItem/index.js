import NewsItem, {
  ItemPropType as _ItemPropType,
} from './component';

export const ItemPropType = _ItemPropType;
export default NewsItem;
