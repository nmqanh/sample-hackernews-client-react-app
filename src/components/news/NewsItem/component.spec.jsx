import React from 'react';
import { advanceTo, clear } from 'jest-date-mock';
import { render } from '@testing-library/react';
import NewsItem from './component';

const sampleNewsItem = {
  objectID: '12345',
  title: 'This is a good news',
  url: 'http://localhost.test',
  points: 20,
  num_comments: 10,
  author: 'Bob',
  created_at: '2020-04-05T03:09:30.000Z',
};

describe('<NewsItem />', () => {
  describe('snapshots', () => {
    beforeAll(() => {
      advanceTo(new Date(2020, 3, 7, 0, 0, 0));
    });

    afterAll(() => {
      clear();
    });

    it('should render NewsItem correctly', () => {
      const { container } = render(
        <NewsItem
          item={sampleNewsItem}
        />,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
