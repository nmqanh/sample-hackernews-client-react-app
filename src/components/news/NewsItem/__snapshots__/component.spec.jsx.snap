// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsItem /> snapshots should render NewsItem correctly 1`] = `
<li
  class="ant-list-item"
>
  <div
    class="ant-list-item-meta"
  >
    <div
      class="ant-list-item-meta-content"
    >
      <h4
        class="ant-list-item-meta-title"
      >
        <a
          href="http://localhost.test"
          target="blank"
        >
          This is a good news
        </a>
      </h4>
      <div
        class="ant-list-item-meta-description"
      >
        <a
          href="http://localhost.test"
          target="blank"
        >
          http://localhost.test
        </a>
      </div>
    </div>
  </div>
  <ul
    class="ant-list-item-action"
  >
    <li>
      <a
        href="https://news.ycombinator.com/item?id=12345"
        style="color: rgba(0, 0, 0, 0.45);"
        target="blank"
      >
        <span
          aria-label="star"
          class="anticon anticon-star"
          role="img"
          style="margin-right: 8px;"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="star"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
            />
          </svg>
        </span>
        20
      </a>
      <em
        class="ant-list-item-action-split"
      />
    </li>
    <li>
      <a
        href="https://news.ycombinator.com/user?id=Bob"
        style="color: rgba(0, 0, 0, 0.45);"
        target="blank"
      >
        <span
          aria-label="user"
          class="anticon anticon-user"
          role="img"
          style="margin-right: 8px;"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="user"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
            />
          </svg>
        </span>
        Bob
      </a>
      <em
        class="ant-list-item-action-split"
      />
    </li>
    <li>
      <a
        href="https://news.ycombinator.com/item?id=12345"
        style="color: rgba(0, 0, 0, 0.45);"
        target="blank"
      >
        <span
          aria-label="clock-circle"
          class="anticon anticon-clock-circle"
          role="img"
          style="margin-right: 8px;"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="clock-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
            <path
              d="M686.7 638.6L544.1 535.5V288c0-4.4-3.6-8-8-8H488c-4.4 0-8 3.6-8 8v275.4c0 2.6 1.2 5 3.3 6.5l165.4 120.6c3.6 2.6 8.6 1.8 11.2-1.7l28.6-39c2.6-3.7 1.8-8.7-1.8-11.2z"
            />
          </svg>
        </span>
        a day ago
      </a>
      <em
        class="ant-list-item-action-split"
      />
    </li>
    <li>
      <a
        href="https://news.ycombinator.com/item?id=12345"
        style="color: rgba(0, 0, 0, 0.45);"
        target="blank"
      >
        <span
          aria-label="message"
          class="anticon anticon-message"
          role="img"
          style="margin-right: 8px;"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="message"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
            />
          </svg>
        </span>
        10
      </a>
    </li>
  </ul>
</li>
`;
